src/namespace.rs

```diff
@@ -0,0 +1,173 @@
+    let (tgid, nstgid) = get_nspid(pid)?;
+    let need_setns = old_stat.ino() != new_stat.ino();
+    let mntns_path = if need_setns {
+        Some(PathBuf::from(new_stat_path))
+    } else {
+        None
+    };
+
+    Ok(NsInfo {
+        tgid,
+        nstgid,
+        need_setns,
+        mntns_path,
+    })
+}
+
+#[cfg(not(target_os = "linux"))]
+pub(crate) fn create_nsinfo(pid: Pid) -> Result<NsInfo, Error> {
+    Ok(NsInfo {
+        tgid: pid,
+        nstgid: pid,
+        need_setns: false,
+        mntns_path: None,
+    })
+}
+
+#[cfg(target_os = "linux")]
+fn get_nspid(pid: Pid) -> Result<(Pid, Pid), Error> {
+    let file = File::open(format!("/proc/{pid}/status"))?;
+    let reader = BufReader::new(file);
+    let (mut tgid, mut nstgid) = (pid, pid);
+    let mut found = false;
+
+    for line in reader.lines() {
+        match line {
+            Ok(line) => {
+                if line.contains("Tgid:") {
+                    if let Some(num) = line
+                        .split("Tgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        tgid = num.into();
+                        nstgid = num.into();
+                        found = true;
+                    }
+                }
+                if line.contains("NStgid:") {
+                    if let Some(num) = line
+                        .split("NStgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        nstgid = num.into();
+                        break;
+                    }
+                }
+            }
+            Err(e) => return Err(e.into()),
+        }
+    }
+
+    if !found {
+        return Err(Error::with_not_found(
+            "failed to get Tgid/NStgid from {file}",
+        ));
+    }
+    Ok((tgid, nstgid))
+}
+
+#[cfg(not(target_os = "linux"))]
+fn get_nspid(pid: Pid) -> Result<(Pid, Pid), Error> {
+    Ok((pid, pid))
+}
+
+impl NsInfo {
+    #[cfg(target_os = "linux")]
+    pub(crate) fn enter_mntns(&self) -> Result<Option<NsCookie>, Error> {
+        if !self.need_setns {
+            return Ok(None);
+        }
+
+        let oldcwd = current_dir()?;
+        let oldns = File::open("/proc/self/ns/mnt")?;
+        let newns = File::open(self.mntns_path.as_ref().unwrap())?;
+        setns(newns, CloneFlags::CLONE_NEWNS).map_err(Into::<std::io::Error>::into)?;
+
+        Ok(Some(NsCookie { oldcwd, oldns }))
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn enter_mntns(&self) -> Result<Option<NsCookie>, Error> {
+        Ok(None)
+    }
+
+    #[cfg(target_os = "linux")]
+    pub(crate) fn exit_mntns(&self, nc: Option<NsCookie>) -> Result<(), Error> {
+        if let Some(nc) = nc {
+            setns(nc.oldns, CloneFlags::CLONE_NEWNS).map_err(Into::<std::io::Error>::into)?;
+            set_current_dir(nc.oldcwd)?;
```

Comment:
I don't really understand why we have to do anything with the current working directory. Is it changed implicitly somewhere?

Reply:
I get the info from https://github.com/torvalds/linux/commit/b01c1f69c8660eaeab7d365cd570103c5c073a02.
But to be honest, I haven't done any testing to verify this (nor have I looked at the implementation of the setns system call to confirm this).

Reply:
Ah, that's why I didn't find it, I was looking at https://github.com/torvalds/linux/commit/843ff37bb59edbe51d64e77ba1b3245a15a4dd9f.

It's dangerous to change the working directory, because we are a library (unlike `perf`) and the working directory is a process-wide thing -- so symbolization has the potential to cause spurious issues in other threads now. But of course if it is changed implicitly by the only API that Linux provides for the purpose we have to accept that... :-(